exo/inference/pytorch/inference.py

```diff
@@ -0,0 +1,336 @@
+# experimental, based off of tinygrad/inference.py
+import asyncio
+import os
+import re
+import numpy as np
+import torch
+import json
+import functools
+from concurrent.futures import ThreadPoolExecutor
+
+from typing import Optional, Tuple, Union, List
+from exo.inference.shard import Shard
+from exo.inference.inference_engine import InferenceEngine
+from exo.inference.pytorch.model.hf import ShardedHuggingFaceModel
+from exo.inference.tokenizers import resolve_tokenizer
+from exo.helpers import DEBUG
+from exo.download.hf.hf_shard_download import HFShardDownloader
+
+from transformers import AutoTokenizer, Cache
+
+# llama
+from transformers.models.llama.modeling_llama import LlamaModel
+
+# model value options
+TOP_K = 20
+TEMP = 0.6
+TOP_P = 0.9
+MAX_LENGTH = 125
+MAX_TIME = 60.0
+
+class PyTorchDynamicShardInferenceEngine(InferenceEngine):
+  """
+  PyTorch Dynamic Shard Inference Engine for performing model inference with sharded Pytorch/HF based models.
+  """
+
+  def __init__(self, shard_downloader: HFShardDownloader):
+    """
+    Initialize the inference engine.
+
+    Args:
+      shard_downloader: Model and weights sharding download
+    """
+    self.shard = None
+    self.shard_downloader = shard_downloader
+
+    # the whole history with new logits need to
+    # be passed to the model to reach the end token
+    # even with caching
+    self.past_input_ids = None
+
+    # setup cuda device
+    if os.environ.get("PYTORCH_DEVICE"):
+      pytorch_device = os.environ["PYTOCH_DEVICE"]
+      if pytorch_device not in ["cuda", "mps"]:
+        pytorch_device = "cpu"
+
+      self.device = pytorch_device
+      self.torch_dtype = torch.float16 if pytorch_device != "cpu" else torch.float32
+
+    if torch.cuda.is_available():
+      self.device = torch.device("cuda")
+      self.torch_dtype = torch.float32
+    elif torch.backends.mps.is_available() and torch.backends.mps.is_built():
+      self.device = torch.device("mps")
+      self.torch_dtype = torch.float32
+    else:
+      self.device = torch.device("cpu")
+      self.torch_dtype = torch.float16
```

Comment:
this looks off:

- the environment variable should override the default behaviour.
- shouldn't `torch.float16 if pytorch_device != "cpu" else torch.float32` be `torch.float16 if pytorch_device == "cpu" else torch.float32`. why is this there in the first place? it seems hacky
- why only support cuda/mps/cpu? what about stuff like `cuda:1` or `xpu`

Reply:
Small addition to Alex's comment above. `pytorch_device = os.environ["PYTOCH_DEVICE"]` has a typo in PYTOCH_DEVICE (should be PYTORCH_DEVICE)

Reply:
Fixing it, silly spelling mistake 🤦🏽‍♂️

Reply:
> this looks off:
> 
> - the environment variable should override the default behaviour.
> - shouldn't `torch.float16 if pytorch_device != "cpu" else torch.float32` be `torch.float16 if pytorch_device == "cpu" else torch.float32`. why is this there in the first place? it seems hacky
> - why only support cuda/mps/cpu? what about stuff like `cuda:1` or `xpu`

True I overcomplicated it. Will fix now

Reply:
Ok I have added new logic for device selection including using the env variable TORCH_DEVICE and using torch.set_default_device and self.device. The reason I just have CPU, CUDA and MPS is because CUDA supports both Nvidia and AMD ROCm (see my [readme](https://github.com/risingsunomi/exo-nvidia/blob/main/exo/inference/torch/README.md)) and to use XLA I would need to install another pytorch library for XLA on a TPU VM.

I also changed the logic for the float32 and float16 selection and cleaned up using the old folder by accident. Hope this works better.